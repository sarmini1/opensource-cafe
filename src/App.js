import React, { useState, useEffect } from "react";
import { BrowserRouter } from "react-router-dom";
import "./App.css";
import SnackOrBoozeApi from "./Api";
import NavBar from "./NavBar";
import RoutesList from "./RoutesList";
import Error from "./Error";

/** App
 *
 * Props:
 * -none
 *
 * State:
 * - isLoading: boolean
 * - snacks: array of objects [{id, name, description, recipe, serve}...]
 * - drinks: array of objects [{id, name, description, recipe, serve}...]
 * - fetchSnacksError: null or ""
 * - fetchDrinksError: null or ""
 *
 * App --> NavBar, Routes
 */
function App() {
  const [isLoading, setIsLoading] = useState(true);
  const [snacks, setSnacks] = useState([]);
  const [drinks, setDrinks] = useState([]);
  const [fetchSnacksError, setFetchSnacksError] = useState(null);
  const [fetchDrinksError, setFetchDrinksError] = useState(null);

  //Grabs all of the snacks after first render, updates snacks state
  useEffect(() => {
    async function fetchSnacks() {
      try {
        let snacks = await SnackOrBoozeApi.getItems("snacks");
        setSnacks(snacks);
        setFetchSnacksError(null);
      } catch (err) {
        console.log(err);
        setFetchSnacksError(err);
      }
    }
    fetchSnacks();
  }, []);

  //Grabs all of the drinks after first render, updates drinks state
  useEffect(() => {
    async function fetchDrinks() {
      try {
        let drinks = await SnackOrBoozeApi.getItems("drinks");
        setDrinks(drinks);
        setIsLoading(false);
        setFetchDrinksError(null);
        /** Only want to set isLoading to false in this effect since it
         * should be the last to run.*/
      } catch (err) {
        setFetchDrinksError(err);
      }
    }
    fetchDrinks();
  }, []);

  // Adds a given item to the appropriate place via an API call
  // and sets the appropriate state, depending on the type of item
  async function addItem(item) {
    // Try/catch will occur when this function is actually called
    let addedItem = await SnackOrBoozeApi.addItem(item);

    if (item.type === "drink") {
      setDrinks(currDrinks => [...currDrinks, addedItem]);
      setIsLoading(false);
    }
    else if (item.type === "snack") {
      setSnacks(currSnacks => [...currSnacks, addedItem]);
      setIsLoading(false);
    }
  }

  if (isLoading) {
    return <p>Loading &hellip;</p>;
  }

  return (
    <div className="App">
      <BrowserRouter>
        <NavBar />
        <main>
          {fetchSnacksError && <Error error={fetchSnacksError} />}
          {fetchDrinksError && <Error error={fetchDrinksError} />}
          <RoutesList
            drinks={drinks}
            snacks={snacks}
            addItem={addItem} />
        </main>
      </BrowserRouter>
    </div>
  );
}

export default App;

/**
 * TODO consider reconfiguring the snacks/drinks state to move
 * into a single piece of state, 'menu', an object with keys for
 * each item group. Each key's value could be an object itself, containing
 * a bunch of different information, including the list of options.
 * Could help with dynamically creating components and reducing headache
 * down the line if other types of items are added to the menu.
 */