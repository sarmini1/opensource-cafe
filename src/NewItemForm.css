/** Add some display, size, padding, and color properties to form components*/
input, select {
  width: 100%;
  padding: 8px 15px;
  margin: 5px;
  display: inline-block;
  border: 1px solid #ccc;
  border-radius: 5px;
}

.btn {
  background-color: #686666;
  margin: 5px;
  color: white;
}

label {
  color: #f7f2f2;
  line-height: 2px;
}