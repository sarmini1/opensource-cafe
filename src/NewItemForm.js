import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import Error from "./Error";
import {
  Card,
  CardBody,
  CardTitle,
  Button,
} from "reactstrap";
import "./NewItemForm.css";

const initialFormData = {
  id: "",
  name: "",
  type: "",
  recipe: "",
  description: "",
  serve: ""
}

/** NewItemForm
 *
 * Props:
 * - addItem: ()
 *
 * State:
 * - NewItemFormData: {}
 * - error: null or ""
 * - isSuccessful: boolean
 *
 * App --> NavBar, Routes --> NewItemForm
 */
function NewItemForm({ addItem }) {

  const [newItemFormData, setNewItemFormData] = useState(initialFormData);
  const [error, setError] = useState(null);
  const navigate = useNavigate();

  //Updates newItemFormData when a change event occurs
  function handleChange(evt) {
    const { name, value } = evt.target;
    setNewItemFormData(currData => ({ ...currData, [name]: value }));
    if (name === "name") {
      setItemID();
    }
  }

  //Adds additional id field to newItemFormData, equal to the
  // item's name, lower case, spaces replaced with dashes
  function setItemID() {
    setNewItemFormData(currData => ({
      ...currData,
      id: currData.name.toLowerCase().replaceAll(" ", "-")
    }));
  }

  //When form is submitted, attempts to add the item via the
  // addItem() function passed from App.
  async function handleSubmit(evt) {
    evt.preventDefault();
    setError(null);

    //Fail fast and check that a valid type was selected
    if (!newItemFormData.type || newItemFormData.type === "default") {
      setError("Please ensure an item type is selected.");
      return;
    }

    try {
      await addItem(newItemFormData);
      setError(null);
      navigate(`/${newItemFormData.type}s`)
    } catch (err) {
      setError(err);
    }
  }

  return (
    <div className="NewItemForm-container">
      <div className="row">
        <div className="col-1 col-sm-2 col-lg-4"></div>
        <div className="col-10 col-sm-8 col-lg-4">
          <Card>
            <CardBody>
              <CardTitle className="font-weight-bold text-center">
                <h3>Add To Our Menu!</h3>
              </CardTitle>
              <form className="NewItemForm-form" onSubmit={handleSubmit}>
                <label htmlFor="NewItemForm-type">
                  Type of Item:
                  </label>
                <select
                  className="NewItemForm-type"
                  name="type"
                  onChange={handleChange}
                >
                  <option value={`default`}>Options:</option>
                  <option value={`snack`}>Snack</option>
                  <option value={`drink`}>Drink</option>
                </select>
                <label htmlFor="NewItemForm-name">
                  Name:
                  </label>
                <input
                  onChange={handleChange}
                  id="NewItemForm-name"
                  name="name"
                  type="text"
                  value={newItemFormData.name}
                  required
                />
                <label htmlFor="NewItemForm-recipe">
                  Recipe:
                  </label>
                <input
                  onChange={handleChange}
                  id="NewItemForm-recipe"
                  name="recipe"
                  type="text"
                  value={newItemFormData.recipe}
                  required
                />
                <label htmlFor="NewItemForm-description">
                  Description:
                  </label>
                <input
                  onChange={handleChange}
                  id="NewItemForm-description"
                  name="description"
                  type="text"
                  value={newItemFormData.description}
                  required
                />
                <label htmlFor="NewItemForm-serving-instructions">
                  Serving Instructions:
                  </label>
                <input
                  onChange={handleChange}
                  id="NewItemForm-serving-instructions"
                  name="serve"
                  type="text"
                  value={newItemFormData.serve}
                  required
                />
                <Button type="submit">Add Item!</Button>
              </form>
              {error && <Error error={error} />}
            </CardBody>
          </Card>
        </div>
        <div className="col-1 col-sm-2 col-lg-4"></div>
      </div>
    </div >
  )

}

export default NewItemForm;