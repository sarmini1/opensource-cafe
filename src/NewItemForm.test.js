import React from 'react';
import { render } from '@testing-library/react';
import NewItemForm from "./NewItemForm";
import { addItem } from "./App";

it('mounts without crashing', function () {
  render(<NewItemForm addItem={addItem} />)
});

it("matches snapshot", function () {
  const { debug, container } = render(<NewItemForm addItem={addItem} />)

  debug(container);
  expect(container).toMatchSnapshot();
});

it("displays a form to add a new item", function () {
  const { debug, container } = render(<NewItemForm addItem={addItem} />);

  debug(container);
  const form = container.querySelector(".NewItemForm-form");
  expect(form).toBeInTheDocument();
});