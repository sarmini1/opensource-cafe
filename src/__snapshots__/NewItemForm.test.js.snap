// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    class="NewItemForm-container"
  >
    <div
      class="row"
    >
      <div
        class="col-1 col-sm-2 col-lg-4"
      />
      <div
        class="col-10 col-sm-8 col-lg-4"
      >
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <div
              class="font-weight-bold text-center card-title"
            >
              <h3>
                Add To Our Menu!
              </h3>
            </div>
            <form
              class="NewItemForm-form"
            >
              <label
                for="NewItemForm-type"
              >
                Type of Item:
              </label>
              <select
                class="NewItemForm-type"
                name="type"
              >
                <option
                  value="default"
                >
                  Options:
                </option>
                <option
                  value="snack"
                >
                  Snack
                </option>
                <option
                  value="drink"
                >
                  Drink
                </option>
              </select>
              <label
                for="NewItemForm-name"
              >
                Name:
              </label>
              <input
                id="NewItemForm-name"
                name="name"
                required=""
                type="text"
                value=""
              />
              <label
                for="NewItemForm-recipe"
              >
                Recipe:
              </label>
              <input
                id="NewItemForm-recipe"
                name="recipe"
                required=""
                type="text"
                value=""
              />
              <label
                for="NewItemForm-description"
              >
                Description:
              </label>
              <input
                id="NewItemForm-description"
                name="description"
                required=""
                type="text"
                value=""
              />
              <label
                for="NewItemForm-serving-instructions"
              >
                Serving Instructions:
              </label>
              <input
                id="NewItemForm-serving-instructions"
                name="serve"
                required=""
                type="text"
                value=""
              />
              <button
                class="btn btn-secondary"
                type="submit"
              >
                Add Item!
              </button>
            </form>
          </div>
        </div>
      </div>
      <div
        class="col-1 col-sm-2 col-lg-4"
      />
    </div>
  </div>
</div>
`;
